### ISSUE TITLE
Older stubs still explicitly using `Any` rather than `Incomplete`

### ISSUE THREAD

**Issue by Avasam** [python/typeshed#9550]
Thanks to typeshed-stats (#9386), I was able to make the following list of stubs majorly using `Any`, indicating that they've likely not been switched over to `Incomplete` yet.
Ordered by ratio of "explicit `Any` / explicit `Incomplete` + 1" parameters and variables.
Excludes stubs marked as obsolete and ratio <= 1 rounded down.

|Package Name|Explicit `Any`|Explicit `Incomplete`|Ratio
|-|-|-|-
|boto|633|0|633
|oauthlib|392|0|392
|passlib|278|0|278
|braintree|506|1|253
|PyYAML|244|0|244
|psycopg2|224|0|224
|Pygments|217|0|217
|html5lib|207|0|207
|stripe|203|0|203
|mock|157|0|157
|protobuf|153|0|153
|commonmark|141|0|141
|babel|134|0|134
|vobject|114|0|114
|beautifulsoup4|103|0|103
|dateparser|101|0|101
|PyMySQL|100|0|100
|Markdown|100|0|100
|httplib2|94|0|94
|aws-xray-sdk|79|0|79
|humanfriendly|76|0|76
|Pillow|397|5|66
|waitress|64|0|64
|invoke|98|1|49
|python-jose|48|0|48
|docutils|47|0|47
|jsonschema|45|0|45
|google-cloud-ndb|259|5|43
|python-gflags|41|0|41
|pyflakes|37|0|37
|setuptools|470|12|36
|mysqlclient|35|0|35
|jmespath|33|0|33
|parsimonious|28|0|28
|opentracing|26|0|26
|regex|25|0|25
|toml|22|0|22
|decorator|22|0|22
|python-xlib|20|0|20
|polib|18|0|18
|requests|64|3|16
|paramiko|46|2|15
|pyserial|14|0|14
|simplejson|14|0|14
|stdlib|2598|186|14
|pynput|12|0|12
|slumber|11|0|11
|console-menu|11|0|11
|SQLAlchemy|2962|301|10
|Flask-Cors|9|0|9
|six|9|0|9
|singledispatch|8|0|8
|caldav|31|3|8
|xmltodict|7|0|7
|paho-mqtt|6|0|6
|redis|549|100|5
|ldap3|557|104|5
|JACK-Client|5|0|5
|typed-ast|5|0|5
|Send2Trash|4|0|4
|pep8-naming|4|0|4
|bleach|4|0|4
|cachetools|4|0|4
|psutil|232|73|3
|python-dateutil|15|4|3
|colorama|3|0|3
|ttkthemes|3|0|3
|ujson|3|0|3
|croniter|3|0|3
|pyvmomi|9|3|2
|flake8-docstrings|2|0|2
|chevron|2|0|2
|contextvars|2|0|2
|first|2|0|2
|retry|2|0|2
|tabulate|2|0|2
|untangle|2|0|2
|toposort|2|0|2
|hdbcli|14|8|2

**Comment by Avasam** [comment]
Do you think it's fine doing a single PR for those (single search & replace change across many third party stubs) instead of 40+ PRs?

Still gotta validate that those with few `Any` actually meant it. Bigger ones can probably be mostly blindly changed?

**Comment by AlexWaygood** [comment]
> Do you think it's fine doing a single PR for those (single search & replace change across many third party stubs) instead of 40-70 PRs?

Maybe we could start off doing a grep for function signatures with `foo: Any | None = ...`. We can be pretty confident that those are artefacts of stubgen, so we can probably update all of those to `foo: Incomplete | None = ...` in a bulk PR pretty safely.

**Comment by srittau** [comment]
I'd say fairly safe, with only few false positives, are:

* Function arguments that contain a union with `Any`. (But not return types.)
* Fields that contain `Any`, either standalone or as union.

It's also not a complete disaster if a wrong `Any` gets changed to `Incomplete`, it just means it needs to be rechecked at some point.

**Comment by Avasam** [comment]
Similar to `: Any | None = ...`, `: Any | None\n` seems like an equivalent stubgen artefact for class variables.

**Comment by AlexWaygood** [comment]
> * Fields that contain `Any`, either standalone or as union.

`: Any | None\n` seems very safe to do a search-and-replace for. For `: Any\n`, I feel a _little_ queasy. But I'd feel more confident if we only replaced places which had, say, three `Any` attributes in a row, e.g.

```python
class Foo:
    a: Any
    b: Any
    c: Any
```

We could find those using regexes or AST.

**Comment by Avasam** [comment]
> Function arguments that contain a union with Any. (But not return types.)

Seems fine to me after running those we deem "safer".

> Fields that contain Any, either standalone [...]

+1 What Alex said.

---

After `: Any | None = ...` and `: Any | None\n` there's 15 `Any | None` left. And except for `mock` they do mostly look like they could be `Incomplete | None`.
- 4 in mock
- 1 in pyOpenSSL
- 1 in redis
- 15 in SQLAlchemy

**Comment by JelleZijlstra** [comment]
It might actually be fine to change every Any in old stubs to Incomplete. In theory, we'll go over the Incompletes later and change them back to Any if appropriate.

**Comment by srittau** [comment]
The only thing I'm a bit uneasy about is changing `Any | None` return types, because they are fairly common workarounds for the missing permissive union type, and they are not that easy to spot.

**Comment by Avasam** [comment]
@srittau 
I agree with return types. I didn't include them in my table above for similar reasons.

TypeAliases should probably also be left untouched if I"m to guess.

**Comment by AlexWaygood** [comment]
> Similar to `: Any | None = ...`, `: Any | None\n` seems like an equivalent stubgen artefact for class variables.

Fancy taking this on as "stage 2", @Avasam? (Regardless of how far we want to go, I think I'd prefer to keep doing this in stages, so we can evaluate the risk level for each stage independently.)

**Comment by Avasam** [comment]
> so we can evaluate the risk level for each stage independently
plus we may learn things so it can be done for stdlib

**Comment by Avasam** [comment]
> The only thing I'm a bit uneasy about is changing `Any | None` return types, because they are fairly common workarounds for the missing permissive union type, and they are not that easy to spot.

We now have `MaybeNone` for those !

**Comment by Avasam** [comment]
I think this ia now well-anchored enough into reviewers and regular contributors' habit that I don't feel the need to track this anymore.
